tools/src/command/prove.rs

```diff
@@ -114,55 +114,181 @@ fn local_prove(
 
+    let mut term = nexus_tui::TerminalHandle::new(TERMINAL_MODE);
+
+    let tr = nexus_api::prover::nova::init_circuit_trace(trace)?;
+    let num_steps = tr.steps();
+
+    let on_step = move |iter: usize| match nova_impl {
+        vm_config::NovaImpl::Parallel | vm_config::NovaImpl::ParallelCompressible => {
+            let b = (num_steps + 1).ilog2();
+            let a = b - 1 - (num_steps - iter).ilog2();
+
+            let step = 2usize.pow(a + 1) * iter - (2usize.pow(a) - 1) * (2usize.pow(b + 1) - 1);
+            let step_type = if iter <= num_steps / 2 {
+                "leaf"
+            } else if iter == num_steps - 1 {
+                "root"
+            } else {
+                "node"
+            };
+            format!("{step_type} {step}")
+        }
+        _ => format!("step {iter}"),
+    };
+
+    let icount = {
+        match nova_impl {
+            vm_config::NovaImpl::Parallel | vm_config::NovaImpl::ParallelCompressible => {
+                k * num_steps
+            }
+            _ => tr.instructions(),
+        }
+    };
+
+    let mut term_ctx = term
+        .context("Computing")
+        .on_step(on_step)
+        .num_steps(num_steps)
+        .with_loading_bar("Proving")
+        .completion_header("Proved")
+        .completion_stats(move |elapsed| {
+            format!(
+                "{num_steps} step(s) in {elapsed}; {:.2} instructions / second",
+                icount as f32 / elapsed.as_secs_f32()
+            )
+        });
+
     match nova_impl {
         vm_config::NovaImpl::Parallel => {
-            let state = nexus_prover::pp::gen_or_load(false, k, path_str, None)?;
-            let root = nexus_prover::prove_par(state, trace)?;
+            let state = {
+                let mut iterm = nexus_tui::TerminalHandle::new(TERMINAL_MODE);
+                let mut term_ctx = iterm
+                    .context("Loading")
+                    .on_step(|_step| "public parameters".into());
+                let _guard = term_ctx.display_step();
+
+                nexus_api::prover::nova::pp::load_pp(path_str)?
+            };
+
+            let mut vs = (0..num_steps)
+                .step_by(2)
+                .map(|i| {
+                    let _guard = term_ctx.display_step();
+
+                    let v = nexus_api::prover::nova::prove_par_leaf_step(&state, &tr, i)?;
+                    Ok(v)
+                })
+                .collect::<anyhow::Result<Vec<_>>>()?;
+
+            let root = {
+                loop {
+                    if vs.len() == 1 {
+                        break;
+                    }
+
+                    vs = vs
+                        .chunks(2)
+                        .map(|ab| {
+                            let _guard = term_ctx.display_step();
+                            let c = nexus_api::prover::nova::prove_par_parent_step(
+                                &state, &tr, &ab[0], &ab[1],
+                            )?;
```

Comment:
Can `vs.len()` ever be odd here? If so, the `ab[1]` indexing will sometimes panic because [slice's `chunks` method](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks) the last chunk returned may be shorter than the specified length when there's a remainder.

Reply:
Ah good call. Used correctly it always is, this is another artifact of losing misuse-resistance in order to move to a step-by-step interface to unhook the telemetry. I'll add an additional check.